refactor the rest package introducing middleware concept and the usage of an http router

Nexucis opened the pull request:
The way to manage the endpoint /metrics is the last thing to do. Otherwise the concepts are in place.

I'm introducing the concept of the middleware which was more or less the case in the previous code. it's just a bit more standard like that with less struct involved.

It could be more standardize if the router used (coming from prometheus) was in a better shape and was allowing the real middleware function. I had to adapt to fit better unfortunately.

It also changed all endpoint available. They are not available through the HTTP method `GET` but through the HTTP method `POST` and all param must be passed through the body instead of using the HTTP query param

Fix #182

slrtbtfs commented:
To me this PR looks quite good and is definitely an improvement over what was there before.

But since I'm not really familiar with the intricacies of writing HTTP API back ends I'd be more comfortable if someone more qualified could review this PR. Maybe @juliusv or @brian-brazil?

---

The contents of `rest/doc.go` should be updated to reflect the changes introduced with this PR.

Nexucis commented:
this PR will impact the way to manage the langServer in the PR https://github.com/prometheus/prometheus/pull/6872

slrtbtfs commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+	}, nil
+}
+
+type API struct {
```
Is this struct intended to satisfy a specific interface?

If yes, a value with the interface type should be returned by `NewLangserver`, to ensure it is actually implemented..

Nexucis replied:
No this struct doesn't implement any kind of interface.
It's just a struct to handle the different endpoints and how they are manages/register

slrtbtfs replied:
Then there should still be a doc comment explaining what this object does.

juliusv replied:
> If yes, a value with the interface type should be returned by NewLangserver, to ensure it is actually implemented..

Btw., even if not applicable here anyway, generally it is encouraged to return concrete implementation types: https://github.com/golang/go/wiki/CodeReviewComments#interfaces, "The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring."

If you want to just test that a type implements an interface, in practice people either just leave that up to the caller / usage site (if that's within the same CI testing realm), or by just saying `var _ MyInterface = &MyImplementation{}` somewhere next to their implementation.

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+
+	err := encoder.Encode(content)
+	if err != nil {
+		http.Error(w, errors.Wrapf(err, "failed to write response").Error(), 500)
```
```suggestion
		http.Error(w, errors.Wrapf(err, "failed to write response").Error(), http.StatusInternalServerError)
```

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+}
+
+type lspData struct {
+	// Expr is the promQL expression and is required for all endpoint available.
```
```suggestion
	// Expr is the PromQL expression and is required for all endpoints available.
```

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+type lspData struct {
+	// Expr is the promQL expression and is required for all endpoint available.
+	Expr string `json:"expr"`
+	// Limit is the number max of result returned to the client. It will be used for the autocompletion and the diagnostics.
```
```suggestion
	// Limit is the maximum number of results returned to the client. It will be used for autocompletion and diagnostics.
```

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+	Limit *uint64 `json:"limit,omitempty"`
+	// PositionLine is the number of the line for which the metadata is queried.
+	PositionLine *float64 `json:"positionLine,omitempty"`
+	// PositionChar for which the metadata is queried. Characters are counted as UTF16 Codepoints.
```
```suggestion
	// PositionChar for which the metadata is queried. Characters are counted as UTF-16 code points.
```

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+	"github.com/prometheus/common/route"
+)
+
+func returnJSON(w http.ResponseWriter, content interface{}) {
```
How about calling this `respondJSON` instead, since that's a clearer description of what it does (not returning a Go value, but sending an HTTP response).

Nexucis replied:
yeah makes sense :)

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+		return err
+	}
+	if len(tmp.Expr) == 0 {
+		return fmt.Errorf("promQL expression is not specified")
```
```suggestion
		return fmt.Errorf("PromQL expression is not specified")
```

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+}
+
+func (d *lspData) UnmarshalJSON(data []byte) error {
+	var tmp lspData
```
Not sure `tmp` is a descriptive name for this. How about just `data` or so?

juliusv replied:
Actually, do you even need `tmp`? I'd just directly try unmarshaling into `d` (cast to `plain`).

Nexucis replied:
yeah you are right. I don't need it

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+	enableMetrics bool
+}
+
+// NewLangServerAPI create a new instance of the Stateless API to use the LangServer through HTTP.
```
```suggestion
// NewLangServerAPI creates a new instance of the Stateless API to use the LangServer through HTTP.
```

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+//
+// The provided Logger should be synchronized.
+//
+// In case "enableMetrics" is set to true, endpoint /metrics is then available and a middleware that instrument the different endpoints provided is instantiated.
```
```suggestion
// In case "enableMetrics" is set to true, endpoint /metrics is then available and a middleware that instruments the different endpoints provided is instantiated.
```

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+// The provided Logger should be synchronized.
+//
+// In case "enableMetrics" is set to true, endpoint /metrics is then available and a middleware that instrument the different endpoints provided is instantiated.
+// Don't use it in case you have already in place such middleware.
```
```suggestion
// Don't use it in case you already have such middleware in place.
```

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+	ctx := r.Context()
+	requestID, requestData, err := getRequestDataAndID(ctx)
+	if err != nil {
+		http.Error(w, err.Error(), 500)
```
Use the `http.StatusXXX` constants here and elsewhere, instead of direct numeric codes (makes it easier for people to immediately understand the meaning of the error).

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+	return nil
+}
+
+func (d *lspData) returnPosition() (protocol.Position, error) {
```
I'd call this just `position()` in Go, returning it is implied...

juliusv commented on rest/doc.go:
```diff
@@ -23,17 +23,19 @@ Supported endpoints:
-	line  : (only for /signatureHelp, /hover and /completion endpoints) The line (0 based) for which the metadata is queried.
-	char  : (only for /signatureHelp, /hover and /completion endpoints) The column (0 based) for which the metadata is queried. Characters are counted as UTF16 Codepoints.
+{
+  "expr": "a promQL expression" # Mandatory for all available endpoint
```
```suggestion
  "expr": "a PromQL expression" # Mandatory for all available endpoints
```

juliusv commented on rest/doc.go:
```diff
@@ -23,17 +23,19 @@ Supported endpoints:
-	char  : (only for /signatureHelp, /hover and /completion endpoints) The column (0 based) for which the metadata is queried. Characters are counted as UTF16 Codepoints.
+{
+  "expr": "a promQL expression" # Mandatory for all available endpoint
+  "limit": 45 # Optional. It will be used only for the endpoint /diagnostics and /completion. It's the maximum number of result returned.
```
```suggestion
  "limit": 45 # Optional. It will be used only for the endpoints /diagnostics and /completion. It's the maximum number of result returned.
```

juliusv commented on rest/doc.go:
```diff
@@ -23,17 +23,19 @@ Supported endpoints:
+  "expr": "a promQL expression" # Mandatory for all available endpoint
+  "limit": 45 # Optional. It will be used only for the endpoint /diagnostics and /completion. It's the maximum number of result returned.
+  "positionLine": 0 # Mandatory for the endpoints /signatureHelp, /hover and /completion. The line (0 based) for which the metadata is queried.
+  "positionChar": 2 # Mandatory for the endpoints /signatureHelp, /hover and /completion. The column (0 based) for which the metadata is queried. Characters are counted as UTF16 Codepoints.
```
```suggestion
  "positionChar": 2 # Mandatory for the endpoints /signatureHelp, /hover and /completion. The column (0 based) for which the metadata is queried. Characters are counted as UTF-16 code points.
```

juliusv commented on rest/middleware.go:
```diff
@@ -0,0 +1,190 @@
+	return "", fmt.Errorf("unable to retrieve the requestID")
+}
+
+// injectRequestData will create a new context and add to this new one the data passed as parameter.
```
```suggestion
// injectRequestData will create a new context and add to this new one to the data passed as parameter.
```

juliusv commented on rest/middleware.go:
```diff
@@ -0,0 +1,190 @@
+func manageDocumentMiddleware(langServer langserver.HeadlessServer) middlewareFunc {
+	return func(next http.HandlerFunc) http.HandlerFunc {
+		return func(w http.ResponseWriter, r *http.Request) {
+			// start to generate an unique ID for the given request
```
```suggestion
			// start to generate a unique ID for the given request
```

Generally in Prometheus we prefer sentence-style comments (start capitalized, end with a period). I don't know about the LSP codebase though.

slrtbtfs replied:
The same is true for the language server, but I'm more lax about it for comments that aren't exported to godoc.

For those that are exported to godoc there's even a linter in place to check it.

juliusv commented on rest/middleware.go:
```diff
@@ -0,0 +1,190 @@
+			// then inject it in the http request context
+			r = r.WithContext(injectRequestID(r.Context(), requestID))
+
+			// then unmarshall the body to the proper struct to be able to retrieve the promQL expr
```
```suggestion
			// then unmarshal the body to the proper struct to be able to retrieve the PromQL expr
```

juliusv commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+	r.Post(prefix+"/hover", a.handle(a.hover))
+	r.Post(prefix+"/signatureHelp", a.handle(a.signature))
+	if a.enableMetrics {
+		r.Get("/metrics", promhttp.Handler().ServeHTTP)
```
It's not common for a type to expose its own metrics via a self-created /metrics endpoint, unless it is really the web handler for your entire application. It would be better IMO to either let the API implement the `prometheus.Collector` interface (like you're currently doing on `apiMetrics`) so that the creator of this API object can register the metrics in whatever existing metrics registry it wants, or to pass in a `prometheus.Registerer` for the same purpose.

Nexucis replied:
in the context of the langServer, the API is actually the web handler for the entire application.
When the API is used outside of the langserver, like in Prometheus, then the metrics implemented is removed because normally the HTTP metrics are implemented by the caller. 

But I think (in another PR), it would make sense to implement other metrics and then exposing them with the way you are proposing

Of course if it's fine with everybody

juliusv replied:
Hm ok, then maybe this can stay like this for now.

juliusv commented:
Thanks! It looks mostly fine, besides my comments.

Nexucis commented:
Thanks for your review @juliusv I fixed all the issue you spotted excepting the last one regarding the way the metrics are exposed. 

Tell me if you are fine with that.

@slrtbtfs I added some code comments regarding the API struct as you was asking :).

On my side, I'm ready to merge it.

slrtbtfs commented on rest/api.go:
```diff
@@ -0,0 +1,243 @@
+	}, nil
+}
+
+// API is the struct that manages the different endpoint provided by the langServer.
```
```suggestion
// API is the struct that manages the different endpoints provided by the language server REST API.
```

slrtbtfs commented on rest/doc.go:
```diff
@@ -23,17 +23,19 @@ Supported endpoints:
-	char  : (only for /signatureHelp, /hover and /completion endpoints) The column (0 based) for which the metadata is queried. Characters are counted as UTF16 Codepoints.
+{
+  "expr": "a PromQL expression" # Mandatory for all available endpoints
+  "limit": 45  # Optional. It will be used only for the endpoints /diagnostics and /completion. It's the maximum number of result returned.
```
```suggestion
  "limit": 45  # Optional. It will be used only for the endpoints /diagnostics and /completion. It's the maximum number of results returned.
```

slrtbtfs commented on rest/doc.go:
```diff
@@ -23,17 +23,19 @@ Supported endpoints:
 	/signatureHelp
 
-URL Query Parameters:
+All endpoint are only available through the HTTP method POST. For each request, you have to provide the following JSON:
```
```suggestion
All endpoints are only available through the HTTP method POST. For each request, you have to provide the following JSON:
```

slrtbtfs commented:
Regrading the `/metrics` endpoint, I'm not even sure myself what specific use case it's aimed at. We might need a bit more clarity about that, before redesigning it.

Nexucis commented:
Thanks for both of you for the english grammatical review ;). Looks like I should take care much more about what I'm writing in english hahaha.

Normally now, everything has been resolved :)
